### ISSUE TITLE
bless to a string ending :: interferes with ISA lookup after typeglob aliasing

### ISSUE THREAD

**Issue by p5pRT** [Perl/perl5#11244]
Migrated from [rt.perl.org#88132](https://rt-archive.perl.org/perl5/Ticket/Display.html?id=88132) (status was 'resolved')

Searchable as RT88132$

**Comment by p5pRT** [comment]
### From @nwc10

#### Created by @nwc10



$ cat package\_corner\_case\.pl 
\#\!/usr/bin/perl \-w
use strict;

my $tail = shift;
$tail = 'inner' unless defined $tail;

@&#8203;left&#8203;::ISA = "outer&#8203;::$tail";
@&#8203;right&#8203;::ISA = "clone&#8203;::$tail";

\*clone&#8203;:: = \*outer&#8203;:: ;

eval "package outer&#8203;::$tail; 1" or die $@&#8203;;

print "ok 1"\, "\\n" if left\->isa\("clone&#8203;::$tail"\);
print "ok 2"\, "\\n" if right\->isa\("outer&#8203;::$tail"\);

print "ok 3"\, "\\n" if right\->isa\("clone&#8203;::$tail"\);
print "ok 4"\, "\\n" if left\->isa\("outer&#8203;::$tail"\);

\_\_END\_\_
$ ~/Sandpit/589/bin/perl5\.8\.9 package\_corner\_case\.pl inner
ok 1
ok 2
ok 3
ok 4
$ ~/Sandpit/589/bin/perl5\.8\.9 package\_corner\_case\.pl inner&#8203;::
ok 1
ok 2
ok 3
ok 4
$ \./perl \-Ilib package\_corner\_case\.pl inner
ok 1
ok 2
ok 3
ok 4
$ \./perl \-Ilib package\_corner\_case\.pl inner&#8203;::
ok 1
ok 3
ok 4


Very obscure\, admittedly\, and probably not anything anyone is relying on\,
but this is a behaviour regression since 5\.8\.x\, and relates to the rather
ugly emergent behaviour of Perl\_gv\_fetch\_pvnflags\(\) as called by
Perl\_gv\_stashpvn\(\)

I don't think that the right fix is "simply to \*copy\*" the logic for
mro\_package\_moved\(\) out to Perl\_gv\_stashpvn\(\)\. The right fix\, to me\, is
changing the parsing logic in Perl\_gv\_fetch\_pvnflags\(\) so that the relevant
code in the single location is always called correctly\.

Nicholas Clark

<details><summary>Perl Info</summary>

```

Flags:
    category=core
    severity=low

Site configuration information for perl 5.14.0:

Configured by nick at Sat Apr  9 10:30:34 BST 2011.

Summary of my perl5 (revision 5 version 14 subversion 0) configuration:
  Derived from: d1ef1aa80747cb421af3e87bd1944a13a5fb2627
  Platform:
    osname=linux, osvers=2.6.35.4, archname=x86_64-linux
    uname='linux eris 2.6.35.4 #4 smp tue sep 21 09:54:22 bst 2010 x86_64 gnulinux '
    config_args='-Dusedevel=y -Dcc=ccache gcc -Dld=gcc -Ubincompat5005 -Uinstallusrbinperl -Dcf_email=nick@ccl4.org -Dperladmin=nick@ccl4.org -Dinc_version_list=  -Dinc_version_list_init=0 -DDEBUGGING -Doptimize=-Os -Uusethreads -Uuselongdouble -Uuse64bitall -Uusemymalloc -Duseperlio -Dprefix=~/Sandpit/snap5.9.x-v5.13.11-407-gd1ef1aa -Uusevendorprefix -Uvendorprefix=~/Sandpit/snap5.9.x-v5.13.11-407-gd1ef1aa -Dinstallman1dir=none -Dinstallman3dir=none -Uuserelocatableinc -Umad -de'
    hint=previous, useposix=true, d_sigaction=define
    useithreads=undef, usemultiplicity=undef
    useperlio=define, d_sfio=undef, uselargefiles=define, usesocks=undef
    use64bitint=define, use64bitall=undef, uselongdouble=undef
    usemymalloc=n, bincompat5005=undef
  Compiler:
    cc='ccache gcc', ccflags ='-DDEBUGGING -fno-strict-aliasing -pipe -fstack-protector -I/usr/local/include -D_LARGEFILE_SOURCE -D_FILE_OFFSET_BITS=64',
    optimize='-g',
    cppflags='-DDEBUGGING -fno-strict-aliasing -pipe -fstack-protector -I/usr/local/include -DDEBUGGING -fno-strict-aliasing -pipe -fstack-protector -I/usr/local/include -D_LARGEFILE_SOURCE -D_FILE_OFFSET_BITS=64 -DDEBUGGING -fno-strict-aliasing -pipe -fstack-protector -I/usr/local/include -D_LARGEFILE_SOURCE -D_FILE_OFFSET_BITS=64 -DDEBUGGING -fno-strict-aliasing -pipe -fstack-protector -I/usr/local/include -D_LARGEFILE_SOURCE -D_FILE_OFFSET_BITS=64 -DDEBUGGING -fno-strict-aliasing -pipe -fstack-protector -I/usr/local/include -D_LARGEFILE_SOURCE -D_FILE_OFFSET_BITS=64 -DDEBUGGING -fno-strict-aliasing -pipe -fstack-protector -I/usr/local/include -D_LARGEFILE_SOURCE -D_FILE_OFFSET_BITS=64 -DDEBUGGING -fno-strict-aliasing -pipe -fstack-protector -I/usr/local/include -D_LARGEFILE_SOURCE -D_FILE_OFFSET_BITS=64 -DDEBUGGING -fno-strict-aliasing -pipe -fstack-protector -I/usr/local/include -D_LARGEFILE_SOURCE -D_FILE_OFFSET_BITS=64 -DDEBUGGING -fno-strict-aliasing -pipe -fstack-protector -I/usr/local/include -D_LARGEFILE_SOURCE -D_FILE_OFFSET_BITS=64 -DDEBUGGING -fno-strict-aliasing -pipe -fstack-protector -I/usr/local/include -D_LARGEFILE_SOURCE -D_FILE_OFFSET_BITS=64 -DDEBUGGING -fno-strict-aliasing -pipe -fstack-protector -I/usr/local/include -D_LARGEFILE_SOURCE -D_FILE_OFFSET_BITS=64'
    ccversion='', gccversion='4.3.2', gccosandvers=''
    intsize=4, longsize=8, ptrsize=8, doublesize=8, byteorder=12345678
    d_longlong=define, longlongsize=8, d_longdbl=define, longdblsize=16
    ivtype='long', ivsize=8, nvtype='double', nvsize=8, Off_t='off_t', lseeksize=8
    alignbytes=8, prototype=define
  Linker and Libraries:
    ld='gcc', ldflags =' -fstack-protector -L/usr/local/lib'
    libpth=/usr/local/lib /lib /usr/lib /lib64 /usr/lib64
    libs=-lnsl -ldb -ldl -lm -lcrypt -lutil -lc
    perllibs=-lnsl -ldl -lm -lcrypt -lutil -lc
    libc=/lib/libc-2.7.so, so=so, useshrplib=false, libperl=libperl.a
    gnulibc_version='2.7'
  Dynamic Linking:
    dlsrc=dl_dlopen.xs, dlext=so, d_dlsymun=undef, ccdlflags='-Wl,-E'
    cccdlflags='-fPIC', lddlflags='-shared -g -L/usr/local/lib -fstack-protector'

Locally applied patches:
    RC0


@INC for perl 5.14.0:
    lib
    /home/nick/Sandpit/snap5.9.x-v5.13.11-379-g801eb08/lib/perl5/site_perl/5.14.0/x86_64-linux
    /home/nick/Sandpit/snap5.9.x-v5.13.11-379-g801eb08/lib/perl5/site_perl/5.14.0
    /home/nick/Sandpit/snap5.9.x-v5.13.11-379-g801eb08/lib/perl5/5.14.0/x86_64-linux
    /home/nick/Sandpit/snap5.9.x-v5.13.11-379-g801eb08/lib/perl5/5.14.0
    .


Environment for perl 5.14.0:
    HOME=/home/nick
    LANG (unset)
    LANGUAGE (unset)
    LD_LIBRARY_PATH (unset)
    LOGDIR (unset)
    PATH=/home/nick/bin:/usr/local/bin:/usr/bin:/bin:/usr/games:/usr/local/sbin:/sbin:/usr/sbin
    PERL_BADLANG (unset)
    SHELL=/bin/bash

```
</details>

**Comment by p5pRT** [comment]
### From @cpansprout

On Sat Apr 09 03&#8203;:59&#8203;:05 2011\, nicholas wrote&#8203;:
> 
> $ cat package\_corner\_case\.pl
> \#\!/usr/bin/perl \-w
> use strict;
> 
> my $tail = shift;
> $tail = 'inner' unless defined $tail;
> 
> @&#8203;left&#8203;::ISA = "outer&#8203;::$tail";
> @&#8203;right&#8203;::ISA = "clone&#8203;::$tail";
> 
> \*clone&#8203;:: = \*outer&#8203;:: ;
> 
> eval "package outer&#8203;::$tail; 1" or die $@&#8203;;
> 
> print "ok 1"\, "\\n" if left\->isa\("clone&#8203;::$tail"\);
> print "ok 2"\, "\\n" if right\->isa\("outer&#8203;::$tail"\);
> 
> print "ok 3"\, "\\n" if right\->isa\("clone&#8203;::$tail"\);
> print "ok 4"\, "\\n" if left\->isa\("outer&#8203;::$tail"\);

That looks awfully familiar\. :\-\)

> 
> \_\_END\_\_
> $ ~/Sandpit/589/bin/perl5\.8\.9 package\_corner\_case\.pl inner
> ok 1
> ok 2
> ok 3
> ok 4
> $ ~/Sandpit/589/bin/perl5\.8\.9 package\_corner\_case\.pl inner&#8203;::
> ok 1
> ok 2
> ok 3
> ok 4
> $ \./perl \-Ilib package\_corner\_case\.pl inner
> ok 1
> ok 2
> ok 3
> ok 4
> $ \./perl \-Ilib package\_corner\_case\.pl inner&#8203;::
> ok 1
> ok 3
> ok 4
> 
> 
> Very obscure\, admittedly\, and probably not anything anyone is relying
>    on\,
> but this is a behaviour regression since 5\.8\.x\, 

I’ve been trying hard to fix all regressions introduced since 5\.8\.8\.
That’s the whole reason I started working on perl\. So I would \*very\*
much like it if this could be made a 5\.14 blocker\. \(I was hoping 5\.14
could be the version that finally fixes all the bugs that 5\.10 introduced\.\) 

> and relates to the
>    rather
> ugly emergent behaviour of Perl\_gv\_fetch\_pvnflags\(\) as called by
> Perl\_gv\_stashpvn\(\)
> 
> I don't think that the right fix is "simply to \*copy\*" the logic for
> mro\_package\_moved\(\) out to Perl\_gv\_stashpvn\(\)\. The right fix\, to me\,
>    is
> changing the parsing logic in Perl\_gv\_fetch\_pvnflags\(\) so that the
>    relevant
> code in the single location is always called correctly\.

Changing the parsing seems the sanest approach\. I think changing

&nbsp;        name\_cursor\+\+;
&nbsp;        name = name\_cursor;

to

&nbsp;        name = name\_cursor \+ 1;

would work\. But I need a day or three to think it through\.

We might instead be able simply to add the name to the stash later in
gv\_fetchpvn\_flags if the full glob name ends with :&#8203;:\. But that might not
fix regressions with regard to \*foo&#8203;::&#8203;::&#8203;:: assignments \(if there are any\)\.

**Comment by p5pRT** [comment]
The RT System itself - Status changed from 'new' to 'open'

**Comment by p5pRT** [comment]
### From @cpansprout

On Sun Apr 10 21&#8203;:55&#8203;:11 2011\, sprout wrote&#8203;:
> On Sat Apr 09 03&#8203;:59&#8203;:05 2011\, nicholas wrote&#8203;:
> > 
> > $ cat package\_corner\_case\.pl
> > \#\!/usr/bin/perl \-w
> > use strict;
> > 
> > my $tail = shift;
> > $tail = 'inner' unless defined $tail;
> > 
> > @&#8203;left&#8203;::ISA = "outer&#8203;::$tail";
> > @&#8203;right&#8203;::ISA = "clone&#8203;::$tail";
> > 
> > \*clone&#8203;:: = \*outer&#8203;:: ;
> > 
> > eval "package outer&#8203;::$tail; 1" or die $@&#8203;;
> > 
> > print "ok 1"\, "\\n" if left\->isa\("clone&#8203;::$tail"\);
> > print "ok 2"\, "\\n" if right\->isa\("outer&#8203;::$tail"\);
> > 
> > print "ok 3"\, "\\n" if right\->isa\("clone&#8203;::$tail"\);
> > print "ok 4"\, "\\n" if left\->isa\("outer&#8203;::$tail"\);
> 
> That looks awfully familiar\. :\-\)
> 
> > 
> > \_\_END\_\_
> > $ ~/Sandpit/589/bin/perl5\.8\.9 package\_corner\_case\.pl inner
> > ok 1
> > ok 2
> > ok 3
> > ok 4
> > $ ~/Sandpit/589/bin/perl5\.8\.9 package\_corner\_case\.pl inner&#8203;::
> > ok 1
> > ok 2
> > ok 3
> > ok 4
> > $ \./perl \-Ilib package\_corner\_case\.pl inner
> > ok 1
> > ok 2
> > ok 3
> > ok 4
> > $ \./perl \-Ilib package\_corner\_case\.pl inner&#8203;::
> > ok 1
> > ok 3
> > ok 4
> > 
> > 
> > Very obscure\, admittedly\, and probably not anything anyone is relying
> >    on\,
> > but this is a behaviour regression since 5\.8\.x\, 
> 
> I’ve been trying hard to fix all regressions introduced since 5\.8\.8\.
> That’s the whole reason I started working on perl\. So I would \*very\*
> much like it if this could be made a 5\.14 blocker\. \(I was hoping 5\.14
> could be the version that finally fixes all the bugs that 5\.10
introduced\.\) 

I forgot to add&#8203;: Since this is such an obscure case\, I think fixing it
is perfectly safe\.

**Comment by p5pRT** [comment]
### From @obra




On Sun\, Apr 10\, 2011 at 09&#8203;:55&#8203;:12PM \-0700\, Father Chrysostomos via RT wrote&#8203;:
> On Sat Apr 09 03&#8203;:59&#8203;:05 2011\, nicholas wrote&#8203;:
> > 
> I???ve been trying hard to fix all regressions introduced since 5\.8\.8\.
> That???s the whole reason I started working on perl\. So I would \*very\*
> much like it if this could be made a 5\.14 blocker\. \(I was hoping 5\.14
> could be the version that finally fixes all the bugs that 5\.10 introduced\.\) 


I'm not sure we'll get all the way to \_that\_ lofty goal for 5\.14\, but
I'm willing to make this one a blocker for now\. Please go ahead :\)

> 
> > and relates to the
> >    rather
> > ugly emergent behaviour of Perl\_gv\_fetch\_pvnflags\(\) as called by
> > Perl\_gv\_stashpvn\(\)
> > 
> > I don't think that the right fix is "simply to \*copy\*" the logic for
> > mro\_package\_moved\(\) out to Perl\_gv\_stashpvn\(\)\. The right fix\, to me\,
> >    is
> > changing the parsing logic in Perl\_gv\_fetch\_pvnflags\(\) so that the
> >    relevant
> > code in the single location is always called correctly\.
> 
> Changing the parsing seems the sanest approach\. I think changing
> 
>         name\_cursor\+\+;
>         name = name\_cursor;
> 
> to
> 
>         name = name\_cursor \+ 1;
> 
> would work\. But I need a day or three to think it through\.
> 
> We might instead be able simply to add the name to the stash later in
> gv\_fetchpvn\_flags if the full glob name ends with :&#8203;:\. But that might not
> fix regressions with regard to \*foo&#8203;::&#8203;::&#8203;:: assignments \(if there are any\)\.
> 
> 

\-\-

**Comment by p5pRT** [comment]
### From @cpansprout

On Sun Apr 10 21&#8203;:55&#8203;:11 2011\, sprout wrote&#8203;:
> Changing the parsing seems the sanest approach\. I think changing
> 
>         name\_cursor\+\+;
>         name = name\_cursor;
> 
> to
> 
>         name = name\_cursor \+ 1;
> 
> would work\.

That seemed the best approach\, except it required a tweak to the
if\(len>0\) condition as well\.

I’ve just applied my change as commit 088225f\.

**Comment by p5pRT** [comment]
@cpansprout - Status changed from 'open' to 'resolved'

**Comment by p5pRT** [comment]
### From @nwc10

On Sun\, Apr 10\, 2011 at 09&#8203;:55&#8203;:12PM \-0700\, Father Chrysostomos via RT wrote&#8203;:
> On Sat Apr 09 03&#8203;:59&#8203;:05 2011\, nicholas wrote&#8203;:

> > print "ok 1"\, "\\n" if left\->isa\("clone&#8203;::$tail"\);
> > print "ok 2"\, "\\n" if right\->isa\("outer&#8203;::$tail"\);
> > 
> > print "ok 3"\, "\\n" if right\->isa\("clone&#8203;::$tail"\);
> > print "ok 4"\, "\\n" if left\->isa\("outer&#8203;::$tail"\);
> 
> That looks awfully familiar\. :\-\)

Yes\, although I did wonder why you'd only output 2 of the 4 possible
combinations in your test programs\. I see that your new commit does all 4\.

> > Very obscure\, admittedly\, and probably not anything anyone is relying
> >    on\,
> > but this is a behaviour regression since 5\.8\.x\, 

> > and relates to the
> >    rather
> > ugly emergent behaviour of Perl\_gv\_fetch\_pvnflags\(\) as called by
> > Perl\_gv\_stashpvn\(\)
> > 
> > I don't think that the right fix is "simply to \*copy\*" the logic for
> > mro\_package\_moved\(\) out to Perl\_gv\_stashpvn\(\)\. The right fix\, to me\,
> >    is
> > changing the parsing logic in Perl\_gv\_fetch\_pvnflags\(\) so that the
> >    relevant
> > code in the single location is always called correctly\.
> 
> Changing the parsing seems the sanest approach\. I think changing
> 
>         name\_cursor\+\+;
>         name = name\_cursor;
> 
> to
> 
>         name = name\_cursor \+ 1;
> 
> would work\. But I need a day or three to think it through\.

That was my roughly thought on the way to go\.

> We might instead be able simply to add the name to the stash later in
> gv\_fetchpvn\_flags if the full glob name ends with :&#8203;:\. But that might not
> fix regressions with regard to \*foo&#8203;::&#8203;::&#8203;:: assignments \(if there are any\)\.

On Wed\, Apr 13\, 2011 at 09&#8203;:51&#8203;:00AM \-0700\, Father Chrysostomos via RT wrote&#8203;:

> That seemed the best approach\, except it required a tweak to the
> if\(len>0\) condition as well\.

but I didn't know that part\.


As you note in 088225fdf76aeaaf's commit message\, it will also change the
structures generated for \*foo&#8203;::&#8203;::&#8203;::\, which was why I asked on list a few
days ago

&nbsp;    Is it documented anywhere how the package names in

&nbsp;    package a&#8203;::&#8203;::&#8203;::b;
&nbsp;    $o = bless \[\]\, "c&#8203;::&#8203;::&#8203;::d";

&nbsp;    are to be parsed?

and everyone who replied stated that there wasn't\.

Nicholas Clark

**Comment by p5pRT** [comment]
### From @cpansprout

On Wed Apr 13 09&#8203;:50&#8203;:59 2011\, sprout wrote&#8203;:
> On Sun Apr 10 21&#8203;:55&#8203;:11 2011\, sprout wrote&#8203;:
> > Changing the parsing seems the sanest approach\. I think changing
> > 
> >         name\_cursor\+\+;
> >         name = name\_cursor;
> > 
> > to
> > 
> >         name = name\_cursor \+ 1;
> > 
> > would work\.
> 
> That seemed the best approach\, except it required a tweak to the
> if\(len>0\) condition as well\.
>
> I’ve just applied my change as commit 088225f\.

That proved to be insufficient\, as the regression still existed in some
cases\. So I have fixed it now \(hopefully\) with commit 1f656fc\.

Since it was intertwined with the bug that has been making RDF&#8203;::Trine
crash \(UNIVERSAL&#8203;::isa\("&#8203;:"\,"foo"\) crashing when \*&#8203;: exists\)\, I have ended
up fixing RDF&#8203;::Trine as well\.
